cups/man/lpinfo.man
.\"
.\" "$Id: lpinfo.man,v 1.1.1.10 2005/01/04 19:16:07 jlovell Exp $"
.\"
.\"   lpinfo man page for the Common UNIX Printing System (CUPS).
.\"
.\"   Copyright 1997-2005 by Easy Software Products.
.\"
.\"   These coded instructions, statements, and computer programs are the
.\"   property of Easy Software Products and are protected by Federal
.\"   copyright law.  Distribution and use rights are outlined in the file
.\"   "LICENSE.txt" which should have been included with this file.  If this
.\"   file is missing or damaged please contact Easy Software Products
.\"   at:
.\"
.\"       Attn: CUPS Licensing Information
.\"       Easy Software Products
.\"       44141 Airport View Drive, Suite 204
.\"       Hollywood, Maryland 20636 USA
.\"
.\"       Voice: (301) 373-9600
.\"       EMail: cups-info@cups.org
.\"         WWW: http://www.cups.org
.\"
.TH lpinfo 8 "Common UNIX Printing System" "23 January 2001" "Easy Software Products"
.SH NAME
lpinfo \- show available devices or drivers
.SH SYNOPSIS
.B lpinfo
[ -E ] [ -l ] [ -m ] [ -v ]
.SH DESCRIPTION
\fBlpinfo\fR lists the available devices or drivers known to the CUPS
server. One of the \fI-m\fR or \fI-v\fR options must be specified to
get any output:
.TP 5
\-E
.br
Forces encryption when connecting to the server.
.TP 5
\-l
.br
Shows a "long" listing of devices or drivers.
.TP 5
\-m
.br
Shows the available printer drivers on the system.
.TP 5
\-v
.br
Shows the available printer devices on the system.
.SH COMPATIBILITY
The \fBlpinfo\fR command is unique to CUPS.
.SH SEE ALSO
lpadmin(8),
CUPS Software Administrators Manual,
http://localhost:631/documentation.html
.SH COPYRIGHT
Copyright 1993-2005 by Easy Software Products, All Rights Reserved.
.\"
.\" End of "$Id: lpinfo.man,v 1.1.1.10 2005/01/04 19:16:07 jlovell Exp $".
.\"


X11/ChangeLog

== 10.4 ChangeLog items ==

2005-03-09  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/darwin.c
	(DarwinUpdateModifiers): handle AlphaShift (caps lock) as a
	normal modifier, now that we disabled XKB (when merging with
	XFree86 4.4) that's the correct thing to do (4015381)

2005-03-01  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: more
	Xpm fixes, from
	https://bugs.freedesktop.org/show_bug.cgi?id=1920, attachment
	1909. (4031381)

	* xc/lib/Xpm/Imakefile: doh, forgot Imakefile portion of
	original patch

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/s_popen.c,
	xc/extras/Xpm/lib/parse.c, xc/extras/Xpm/lib/misc.c,
	xc/extras/Xpm/lib/hashtab.c, xc/extras/Xpm/lib/data.c,
	xc/extras/Xpm/lib/create.c, xc/extras/Xpm/lib/XpmI.h,
	xc/extras/Xpm/lib/WrFFrI.c, xc/extras/Xpm/lib/WrFFrBuf.c,
	xc/extras/Xpm/lib/RdFToI.c, xc/extras/Xpm/lib/RdFToBuf.c,
	xc/extras/Xpm/lib/Imakefile, xc/extras/Xpm/lib/CrDatFrI.c,
	xc/extras/Xpm/lib/CrBufFrI.c, xc/extras/Xpm/lib/Attrib.c:
	merged changes from XFree86 HEAD branch (4022757)

2005-02-15  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/quartz-server.c: request that
	deferred updates are disabled. Otherwise we lose 25%
	performance on the "cat file in xterm" test (3985081)

	Fix 3494176 (deadlock while closing OpenGL window):

	* xc/lib/GL/glx/glxext.c: make __glXLock a recursive mutex

	* xc/lib/GL/apple/dri_driver.h, xc/lib/GL/apple/dri_driver.c:
	make the screen mutex a pthread_mutex and make it recursive

2005-01-04  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/X11Controller.m: fixed several
	bugs in the customization dialog for the applications menu
	(3668769)

2004-11-16  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: added a signal
	handler to catch SIGTERM etc and exit cleanly, killing the X
	server and xinit (3879878)

2004-10-25  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/English.lproj/InfoPlist.strings:
	removed CFBundleShortVersionString and CFBundleGetInfoString
	(3851089)

2004-10-19  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/English.lproj/Localizable.strings:
	added "Terminal" to the strings file	

	* xc/programs/Xserver/hw/apple/X11Application.m (prefs_get):
	added code to call NSLocalizedString on each name in the
	default applications menu

	* xc/programs/Xserver/hw/apple/rootless-window.c
	(RootlessKnowsWindowNumber): added check for zero window id

2004-10-15  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/English.lproj/Localizable.strings:
	Expanded backslash escape sequences since .strings files don't
	seem to handle them correctly. (3833239)

2004-10-15  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/vfb/InitOutput.c: workaround 3834723
	by rearranging include order

2004-10-10  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/quartz.h,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/X11Controller.m: QuartzRunClient
	function now has a second arg - whether or not to quote the
	command (3816789, 3572554)

2004-10-08  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/parse.c,
	xc/extras/Xpm/lib/hashtab.c, xc/extras/Xpm/lib/data.c,
	xc/extras/Xpm/lib/create.c, xc/extras/Xpm/lib/XpmI.h,
	xc/extras/Xpm/lib/WrFFrI.c, xc/extras/Xpm/lib/CrDatFrI.c,
	xc/extras/Xpm/lib/Attrib.c: merged Xpm changes from XFree86
	HEAD branch (3744542, 3778302)

2004-09-29  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/English.lproj/main.nib/: fixed
	typo (3820725)

2004-09-16  John Harper  <jsh@apple.com>

	* xc/programs/xterm/main.c: call openpty() to scan for a pseudo
	terminal (3768572)

	* xc/programs/xterm/xterm.man: fix documentation of -ls option
	(3803120)

2004-09-08  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/English.lproj/main.nib/: nib
	changes for 3313133 and 3316468

2004-09-03  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/English.lproj/main.nib/: gave
	applications dialog window a minimum size

	* xc/programs/Xserver/hw/apple/English.lproj/Localizable.strings:
	new file - ran genstrings on the .m files

	* xc/programs/Xserver/hw/apple/X11.pbproj/project.pbxproj:
	added Localizable.strings

	* xc/programs/Xserver/hw/apple/X11Controller.m,
	xc/programs/Xserver/hw/apple/X11Application.m: added comments
	for translators

	* xc/programs/Xserver/hw/apple/Imakefile: set OBJROOT and
	SYMROOT when calling xcodebuild

2004-08-25  John Harper  <jsh@apple.com>

	* quartz-wm/x-window.m: don't grab Mod1-grave for window
	cycling, only Mod2 (3774967)

2004-08-23  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: fix typo in
	$XINITRC code (3621248)

2004-07-16  John Harper  <jsh@apple.com>

	* merged changes from XFree86 4.4.0 and fixed resulting
	brokenness (3570235, 3612446, 3548675, 3404805)

2004-06-10  John Harper  <jsh@apple.com>

	* xinitrc: quote shell expanded filenames to support home
	directories with spaces (3674992)

2004-03-29  John Harper  <jsh@apple.com>

	* FS/FSFontInfo.c, FS/FSFtNames.c, FS/FSGetCats.c,
	FS/FSListCats.c, FS/FSListExt.c, FS/FSOpenServ.c,
	FS/FSQGlyphs.c, FS/FSQXExt.c, FS/FSQXInfo.c, FS/FSlibint.h,
	FS/FSlibos.h, font/fc/fsconvert.c, font/fc/fserve.c,
	font/fc/fslibos.h, font/fontfile/dirfile.c,
	font/fontfile/encparse.c, font/fontfile/fontdir.c,
	font/fontfile/fontfile.c: merged font security fixes from
	XFree86 4.3 branch

2004-03-11  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/quartz.c (QuartzRunClient):
	added command quoting hack to fix 3572554

	* xc/programs/Xserver/hw/apple/bundle-main.c (define_named):
	fix for 3571829 (writing bad address size into xauth item)

2004-02-09  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c (start_client):
	respect value of $XINITRC. Also default to xinitrc in
	/etc/X11/xinit not the one under /usr/X11R6


cups/systemv/lpinfo.c
/*
 * "$Id: lpinfo.c,v 1.1.1.11 2005/01/04 19:16:40 jlovell Exp $"
 *
 *   "lpinfo" command for the Common UNIX Printing System (CUPS).
 *
 *   Copyright 1997-2005 by Easy Software Products.
 *
 *   These coded instructions, statements, and computer programs are the
 *   property of Easy Software Products and are protected by Federal
 *   copyright law.  Distribution and use rights are outlined in the file
 *   "LICENSE.txt" which should have been included with this file.  If this
 *   file is missing or damaged please contact Easy Software Products
 *   at:
 *
 *       Attn: CUPS Licensing Information
 *       Easy Software Products
 *       44141 Airport View Drive, Suite 204
 *       Hollywood, Maryland 20636 USA
 *
 *       Voice: (301) 373-9600
 *       EMail: cups-info@cups.org
 *         WWW: http://www.cups.org
 *
 * Contents:
 *
 *   main()         - Parse options and show information.
 *   show_devices() - Show available devices.
 *   show_models()  - Show available PPDs.
 */

/*
 * Include necessary headers...
 */

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <cups/cups.h>
#include <cups/language.h>
#include <cups/debug.h>
#include <cups/string.h>


/*
 * Local functions...
 */

static int	show_devices(http_t *, int);
static int	show_models(http_t *, int);


/*
 * 'main()' - Parse options and show status information.
 */

int
main(int  argc,			/* I - Number of command-line arguments */
     char *argv[])		/* I - Command-line arguments */
{
  int		i;		/* Looping var */
  http_t	*http;		/* Connection to server */
  int		long_status;	/* Long listing? */


  http        = NULL;
  long_status = 0;

  for (i = 1; i < argc; i ++)
    if (argv[i][0] == '-')
      switch (argv[i][1])
      {
        case 'E' : /* Encrypt */
#ifdef HAVE_SSL
	    cupsSetEncryption(HTTP_ENCRYPT_REQUIRED);

	    if (http)
	      httpEncryption(http, HTTP_ENCRYPT_REQUIRED);
#else
            fprintf(stderr, "%s: Sorry, no encryption support compiled in!\n",
	            argv[0]);
#endif /* HAVE_SSL */
	    break;

        case 'l' : /* Show long listing */
	    long_status = 1;
	    break;

        case 'm' : /* Show models */
	    if (!http)
	    {
              http = httpConnectEncrypt(cupsServer(), ippPort(),
	                                cupsEncryption());

	      if (http == NULL)
	      {
		perror("lpinfo: Unable to connect to server");
		return (1);
	      }
            }

            if (show_models(http, long_status))
	      return (1);
	    break;
	    
        case 'v' : /* Show available devices */
	    if (!http)
	    {
              http = httpConnectEncrypt(cupsServer(), ippPort(),
	                                cupsEncryption());

	      if (http == NULL)
	      {
		perror("lpinfo: Unable to connect to server");
		return (1);
	      }
            }

            if (show_devices(http, long_status))
	      return (1);
	    break;

        case 'h' : /* Connect to host */
	    if (http)
	    {
	      httpClose(http);
	      http = NULL;
	    }

	    if (argv[i][2] != '\0')
	      cupsSetServer(argv[i] + 2);
	    else
	    {
	      i ++;

	      if (i >= argc)
	      {
	        fputs("Error: need hostname after \'-h\' option!\n", stderr);
		return (1);
              }

	      cupsSetServer(argv[i]);
	    }
	    break;

	default :
	    fprintf(stderr, "lpinfo: Unknown option \'%c\'!\n", argv[i][1]);
	    return (1);
      }
    else
    {
      fprintf(stderr, "lpinfo: Unknown argument \'%s\'!\n", argv[i]);
      return (1);
    }

  return (0);
}


/*
 * 'show_devices()' - Show available devices.
 */

static int			/* O - 0 on success, 1 on failure */
show_devices(http_t *http,	/* I - HTTP connection to server */
             int    long_status)/* I - Long status report? */
{
  ipp_t		*request,	/* IPP Request */
		*response;	/* IPP Response */
  ipp_attribute_t *attr;	/* Current attribute */
  cups_lang_t	*language;	/* Default language */
  const char	*device_class,	/* Pointer into device-class */
		*device_info,	/* Pointer into device-info */
		*device_make,	/* Pointer into device-make-and-model */
		*device_uri;	/* Pointer into device-uri */


  if (http == NULL)
    return (1);

 /*
  * Build a CUPS_GET_DEVICES request, which requires the following
  * attributes:
  *
  *    attributes-charset
  *    attributes-natural-language
  *    printer-uri
  */

  request = ippNew();

  request->request.op.operation_id = CUPS_GET_DEVICES;
  request->request.op.request_id   = 1;

  language = cupsLangDefault();

  ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_CHARSET,
               "attributes-charset", NULL, cupsLangEncoding(language));

  ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_LANGUAGE,
               "attributes-natural-language", NULL, language->language);

  ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_URI, "printer-uri",
               NULL, "ipp://localhost/printers/");

 /*
  * Do the request and get back a response...
  */

  if ((response = cupsDoRequest(http, request, "/")) != NULL)
  {
   /*
    * Loop through the device list and display them...
    */

    if (response->request.status.status_code > IPP_OK_CONFLICT)
    {
      fprintf(stderr, "lpinfo: cups-get-devices failed: %s\n",
              ippErrorString(response->request.status.status_code));
      ippDelete(response);
      return (1);
    }

    for (attr = response->attrs; attr != NULL; attr = attr->next)
    {
     /*
      * Skip leading attributes until we hit a device...
      */

      while (attr != NULL && attr->group_tag != IPP_TAG_PRINTER)
        attr = attr->next;

      if (attr == NULL)
        break;

     /*
      * Pull the needed attributes from this device...
      */

      device_class = NULL;
      device_info  = NULL;
      device_make  = NULL;
      device_uri   = NULL;

      while (attr != NULL && attr->group_tag == IPP_TAG_PRINTER)
      {
        if (strcmp(attr->name, "device-class") == 0 &&
	    attr->value_tag == IPP_TAG_KEYWORD)
	  device_class = attr->values[0].string.text;

        if (strcmp(attr->name, "device-info") == 0 &&
	    attr->value_tag == IPP_TAG_TEXT)
	  device_info = attr->values[0].string.text;

        if (strcmp(attr->name, "device-make-and-model") == 0 &&
	    attr->value_tag == IPP_TAG_TEXT)
	  device_make = attr->values[0].string.text;

        if (strcmp(attr->name, "device-uri") == 0 &&
	    attr->value_tag == IPP_TAG_URI)
	  device_uri = attr->values[0].string.text;

        attr = attr->next;
      }

     /*
      * See if we have everything needed...
      */

      if (device_class == NULL || device_info == NULL ||
          device_make == NULL || device_uri == NULL)
      {
        if (attr == NULL)
	  break;
	else
          continue;
      }

     /*
      * Display the device...
      */

      if (long_status)
      {
	printf("Device: uri = %s\n", device_uri);
	printf("        class = %s\n", device_class);
	printf("        info = %s\n", device_info);
	printf("        make-and-model = %s\n", device_make);
      }
      else
        printf("%s %s\n", device_class, device_uri);

      if (attr == NULL)
        break;
    }

    ippDelete(response);
  }
  else
  {
    fprintf(stderr, "lpinfo: cups-get-devices failed: %s\n",
            ippErrorString(cupsLastError()));
    return (1);
  }

  return (0);
}


/*
 * 'show_models()' - Show available PPDs.
 */

static int			/* O - 0 on success, 1 on failure */
show_models(http_t *http,	/* I - HTTP connection to server */
            int    long_status)	/* I - Long status report? */
{
  ipp_t		*request,	/* IPP Request */
		*response;	/* IPP Response */
  ipp_attribute_t *attr;	/* Current attribute */
  cups_lang_t	*language;	/* Default language */
  const char	*ppd_language,	/* Pointer into ppd-natural-language */
		*ppd_make,	/* Pointer into ppd-make-and-model */
		*ppd_name;	/* Pointer into ppd-name */


  if (http == NULL)
    return (1);

 /*
  * Build a CUPS_GET_PPDS request, which requires the following
  * attributes:
  *
  *    attributes-charset
  *    attributes-natural-language
  *    printer-uri
  */

  request = ippNew();

  request->request.op.operation_id = CUPS_GET_PPDS;
  request->request.op.request_id   = 1;

  language = cupsLangDefault();

  ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_CHARSET,
               "attributes-charset", NULL, cupsLangEncoding(language));

  ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_LANGUAGE,
               "attributes-natural-language", NULL, language->language);

  ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_URI, "printer-uri",
               NULL, "ipp://localhost/printers/");

 /*
  * Do the request and get back a response...
  */

  if ((response = cupsDoRequest(http, request, "/")) != NULL)
  {
   /*
    * Loop through the device list and display them...
    */

    if (response->request.status.status_code > IPP_OK_CONFLICT)
    {
      fprintf(stderr, "lpinfo: cups-get-ppds failed: %s\n",
              ippErrorString(response->request.status.status_code));
      ippDelete(response);
      return (1);
    }

    for (attr = response->attrs; attr != NULL; attr = attr->next)
    {
     /*
      * Skip leading attributes until we hit a PPD...
      */

      while (attr != NULL && attr->group_tag != IPP_TAG_PRINTER)
        attr = attr->next;

      if (attr == NULL)
        break;

     /*
      * Pull the needed attributes from this PPD...
      */

      ppd_language = NULL;
      ppd_make     = NULL;
      ppd_name     = NULL;

      while (attr != NULL && attr->group_tag == IPP_TAG_PRINTER)
      {
        if (strcmp(attr->name, "ppd-natural-language") == 0 &&
	    attr->value_tag == IPP_TAG_LANGUAGE)
	  ppd_language = attr->values[0].string.text;

        if (strcmp(attr->name, "ppd-make-and-model") == 0 &&
	    attr->value_tag == IPP_TAG_TEXT)
	  ppd_make = attr->values[0].string.text;

        if (strcmp(attr->name, "ppd-name") == 0 &&
	    attr->value_tag == IPP_TAG_NAME)
	  ppd_name = attr->values[0].string.text;

        attr = attr->next;
      }

     /*
      * See if we have everything needed...
      */

      if (ppd_language == NULL || ppd_make == NULL || ppd_name == NULL)
      {
        if (attr == NULL)
	  break;
	else
          continue;
      }

     /*
      * Display the device...
      */

      if (long_status)
      {
	printf("Model:  name = %s\n", ppd_name);
	printf("        natural_language = %s\n", ppd_language);
	printf("        make-and-model = %s\n", ppd_make);
      }
      else
        printf("%s %s\n", ppd_name, ppd_make);

      if (attr == NULL)
        break;
    }

    ippDelete(response);
  }
  else
  {
    fprintf(stderr, "lpinfo: cups-get-ppds failed: %s\n",
            ippErrorString(cupsLastError()));

    return (1);
  }

  return (0);
}


/*
 * End of "$Id: lpinfo.c,v 1.1.1.11 2005/01/04 19:16:40 jlovell Exp $".
 */
